[MRG] documentation for random_state in forest.py

MDouriez opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
part of this issue: https://github.com/scikit-learn/scikit-learn/issues/15222
(related to https://github.com/scikit-learn/scikit-learn/pull/15264)


#### What does this implement/fix? Explain your changes.
This PR improves the documentation for `RandomForestClassifier`, `RandomForestRegressor`, `ExtraTreesClassifier` and `ExtraTreesRegressor` regarding the source of randomness. The only changes were made in the docstrings of `sklearn/ensemble/_forest.py`.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

MDouriez committed:
documentation for random_state in forests

eickenberg commented:
Looks good to me pending CI

MDouriez committed:
move note to parameter

MDouriez committed:
same for RandomForestRegressor

MDouriez committed:
add doc for ExtraTreesRegressor and ExtraTreesClassifier

MDouriez commented:
@NicolasHug Let me know what you think. Thanks!

cmarmo commented:
Also @TomDLT or maybe @adrinjalali (seems to me that you volunteered  to be pinged... ;) )... this PR looks pretty ready for merging?

adrinjalali commented:
Thanks for the ping @cmarmo 

I like the description, but I'm not sure if the part which is related to the randomness in trees should be included here or not. What do you think of adding a link to the tree docstring intead for that part @MDouriez ?

NicolasHug commented on sklearn/ensemble/_forest.py:
```diff
@@ -960,10 +960,17 @@ class RandomForestClassifier(ForestClassifier):
+        data, ``max_features=n_features`` and ``bootstrap=False``, if the
+        improvement of the criterion is identical for several splits enumerated
+        during the search of the best split. To obtain a deterministic
+        behaviour during fitting, ``random_state`` has to be fixed.
```
I think we can omit this last part (from "Also note that..." to "has to be fixed").

The rest above is perfect

MDouriez replied:
This note was already there. I just moved it. Should I still remove it?

NicolasHug replied:
I see, maybe just leave it where it was in the notes then

NicolasHug commented on sklearn/ensemble/_forest.py:
```diff
@@ -1552,10 +1552,13 @@ class ExtraTreesClassifier(ForestClassifier):
-        If RandomState instance, random_state is the random number generator;
-        If None, the random number generator is the RandomState instance used
-        by `np.random`.
+        Controls 3 sources of randomness:
```
Rendering is a bit weird, I think you need to skip a line after this one


see https://81655-843222-gh.circle-artifacts.com/0/doc/modules/generated/sklearn.ensemble.ExtraTreesClassifier.html

MDouriez replied:
Thank you Nicolas. How did you generate the preview?

NicolasHug replied:
https://scikit-learn.org/dev/developers/contributing.html#generated-documentation-on-circleci

but you can of course also build the docs locally

NicolasHug reviewed (commented):
This is a clear improvement, thanks @MDouriez . Made some comments but LGTM when addressed

Commit:
skip line

Commit:
lint

MDouriez committed:
move note back to where it was

MDouriez committed:
add Glossary in RandomForestRegressor

MDouriez committed:
adding description for RandomTreesEmbedding

MDouriez committed:
Merge branch 'master' into 15222_forest

MDouriez committed:
small fix

MDouriez commented:
Should be ready for final review @NicolasHug @adrinjalali.
- Left the note where it was
- Added line for rendering
- Added description for `RandomTreesEmbedding` as well

MDouriez committed:
correct description for RandomTreesEmbedding

MDouriez commented:
Also interestingly, `RandomTreesEmbedding` has a `max_samples` parameter but no `bootstrap`. In the `__init__`, `bootstrap` is set to `False`. Looks like `max_samples` is never used?

NicolasHug reviewed (approved):
Thanks @MDouriez !

NicolasHug commented:
> Also interestingly, RandomTreesEmbedding has a max_samples parameter but no bootstrap. In the __init__, bootstrap is set to False. Looks like max_samples is never used?

Good catch, could you please open an issue for this?

adrinjalali reviewed (approved):
Thanks @MDouriez

MDouriez commented:
filed an issue https://github.com/scikit-learn/scikit-learn/issues/15670

> > Also interestingly, RandomTreesEmbedding has a max_samples parameter but no bootstrap. In the **init**, bootstrap is set to False. Looks like max_samples is never used?
> 
> Good catch, could you please open an issue for this?
